quickwit/quickwit-ingest/src/ingest_v2/ingester.rs

```diff
@@ -0,0 +1,1257 @@
+
+        let mut ingester = self.ingester_pool.get(follower_id).await.ok_or(
+            IngestV2Error::IngesterUnavailable {
+                ingester_id: follower_id.clone(),
+            },
+        )?;
+        let mut ack_replication_stream = ingester
+            .open_replication_stream(syn_replication_stream)
+            .await?;
+        ack_replication_stream
+            .next()
+            .await
+            .expect("FIXME")
+            .expect("")
+            .into_open_response()
+            .expect("The first message should be an open response.");
+
+        let replication_client =
+            ReplicationClientTask::spawn(syn_replication_stream_tx, ack_replication_stream);
+        entry.insert(replication_client);
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl IngesterService for Ingester {
+    async fn persist(
+        &mut self,
+        persist_request: PersistRequest,
+    ) -> IngestV2Result<PersistResponse> {
+        if persist_request.leader_id != self.self_node_id {
+            return Err(IngestV2Error::Internal(format!(
+                "Routing error: request was sent to ingester node `{}` instead of `{}`.",
+                self.self_node_id, persist_request.leader_id,
+            )));
+        }
+        let mut state_guard = self.state.write().await;
+
+        let mut persist_successes = Vec::with_capacity(persist_request.subrequests.len());
+        // let mut persist_failures = Vec::new();
+        let mut replicate_subrequests: HashMap<NodeId, Vec<ReplicateSubrequest>> = HashMap::new();
+
+        let commit_type = persist_request.commit_type();
+        let force_commit = commit_type == CommitTypeV2::Force;
+        let leader_id: NodeId = persist_request.leader_id.into();
+
+        for subrequest in persist_request.subrequests {
+            let queue_id = subrequest.queue_id();
+            let follower_id: Option<NodeId> = subrequest.follower_id.map(Into::into);
+            let primary_shard =
+                if let Some(primary_shard) = state_guard.primary_shards.get(&queue_id) {
+                    primary_shard
+                } else {
+                    self.init_primary_shard(
+                        &mut state_guard,
+                        &queue_id,
+                        &leader_id,
+                        follower_id.as_ref(),
+                    )
+                    .await?
+                };
+            let from_position_inclusive = primary_shard.primary_position_inclusive;
+
+            let Some(doc_batch) = subrequest.doc_batch else {
+                let persist_success = PersistSuccess {
+                    index_uid: subrequest.index_uid,
+                    source_id: subrequest.source_id,
+                    shard_id: subrequest.shard_id,
+                    replication_position_inclusive: from_position_inclusive.offset(),
+                };
+                persist_successes.push(persist_success);
```

Comment:
could we have a `force_commit` request with no batch?

Reply:
This could be useful. An empty batch without force commit should be no-op. An empty batch with force commit should trigger a commit.

I don't think I have implemented this behavior. Will add it to the to-do list.